rules/gcs.bzl

```diff
@@ -0,0 +1,91 @@
+# Copyright 2017 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Skylark rule for pulling a file from GCS bucket.
+"""
+
+_GCS_FILE_BUILD = """
+package(default_visibility = ["//visibility:public"])
+filegroup(
+    name = "file",
+    srcs = ["{}"],
+)
+"""
+
+def _gcs_file_impl(ctx):
+    """Implementation of the gcs_file rule."""
+    repo_root = ctx.path(".")
+    forbidden_files = [
+        repo_root,
+        ctx.path("WORKSPACE"),
+        ctx.path("BUILD"),
+        ctx.path("BUILD.bazel"),
+        ctx.path("file/BUILD"),
+        ctx.path("file/BUILD.bazel"),
+    ]
+    downloaded_file_path = ctx.attr.downloaded_file_path or ctx.attr.file
+    download_path = ctx.path("file/" + downloaded_file_path)
+    if download_path in forbidden_files or not str(download_path).startswith(str(repo_root)):
+        fail("'%s' cannot be used as downloaded_file_path in gcs_file" % ctx.attr.downloaded_file_path)
+
+    # Add a top-level BUILD file to export all the downloaded files.
+    ctx.file("file/BUILD", _GCS_FILE_BUILD.format(downloaded_file_path))
```

Comment:
AFAIK quite some different rules would use the folder `file`. Since we are dealing with GCS file here so let's be more specific and call it `gcsfile` and the build file `gcsfile/BUILD`

Reply:
this should be fine as file. This repo is created by this rule, and no other rules should create any files under this folder.